### PR TITLE
Kionix:KX022 driver

### PR DESCRIPTION
**Pull request by AzamLukman**
Addi Kionix KX022 Driver

### CODE REVIEW

File: drivers/sensor/kx022/kx022_trigger.c [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+	struct kx022_data *data = dev->data;
+	const struct kx022_config *cfg = dev->config;
+	uint8_t status, clr;
+
+	if (data->hw_tf->read_reg(data, KX022_REG_INS2, &status) < 0) {
+		return;
+	}
+
+	if (status & 0x02) {
+		printf("===================MOTION DETECTED===============\r\n");
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:70]: Have to remove these or print them with `LOG_DBG` or else there will be an impact to the performance.

File: drivers/sensor/kx022/kx022_trigger.c [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+	}
+}
+
+static void kx022_handle_int(const struct device *dev)
+{
+	struct kx022_data *data = dev->data;
+	const struct kx022_config *cfg = dev->config;
+	uint8_t status, clr;
+
+	if (data->hw_tf->read_reg(data, KX022_REG_INS2, &status) < 0) {
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:65]: Unless only INT2 is supported, if INT1 / INT2 are configurable in Kconfig then you will have to handle the possibilities of both cases here?

File: drivers/sensor/kx022/kx022_trigger.c [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+	if (status & 0x02) {
+		printf("===================MOTION DETECTED===============\r\n");
+		kx022_handle_motion_int(dev);
+	} else if (status & 0x10) {
+		printf("======================DATA READY==================\r\n");
+
+		kx022_handle_drdy_int(dev);
+		k_msleep(50);
+
+	} else if (status & 0x01) {
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:78]: https://github.com/RohmSemiconductor/RegisterDefinitions/blob/master/kx022_registers.h#L169

image

File: drivers/sensor/kx022/kx022_trigger.c [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+}
+#endif
+
+/************************************************************
+ * info: kx022_mode use to set pc1 in standby or operating mode
+ * input: 0 - standby
+ *        1 - operating
+ * **********************************************************/
+
+void kx022_mode(const struct device *dev, bool mode)
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:111]: This should be in the main `kx022.c` file. If user chose not to compile the trigger functions this won't be compiled altogether and would result in compile error.

File: drivers/sensor/kx022/kx022_trigger.c [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+
+	gpio_init_callback(&data->gpio_cb, kx022_gpio_callback, BIT(cfg->irq_pin));
+
+	if (gpio_add_callback(data->gpio, &data->gpio_cb) < 0) {
+		LOG_ERR("Could not set gpio callback.");
+		return -EIO;
+	}
+	data->dev = dev;
+	switch (KX022_DEFAULT_INT_PIN) {
+	case 1:
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:145]: The defined values in the header file can be handy

File: drivers/sensor/kx022/kx022_trigger.c [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+		data->hw_tf->update_reg(data, KX022_REG_INC1, KX022_MAKS_INC1_INT_EN,
+					KX022_INT1_EN);
+		break;
+	case 2:
+		data->hw_tf->update_reg(data, KX022_REG_INC5, KX022_MAKS_INT2_EN, KX022_INT2_EN);
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:150]: Not quite symmetric with its INT1 counterpart?

File: drivers/sensor/kx022/kx022_trigger.c [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+		break;
+	default:
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:152]: default case likely not required here, if there's only 2 possibilities in the Kconfig

File: drivers/sensor/kx022/kx022_trigger.c [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+
+#if defined(CONFIG_KX022_TRIGGER_OWN_THREAD)
+	k_sem_init(&data->trig_sem, 0, UINT_MAX);
+
+	k_thread_create(&data->thread, data->thread_stack, CONFIG_KX022_THREAD_STACK_SIZE,
+			(k_thread_entry_t)kx022_thread, data, NULL, NULL,
+			K_PRIO_COOP(CONFIG_KX022_THREAD_PRIORITY), 0, K_NO_WAIT);
+#elif defined(CONFIG_KX022_TRIGGER_GLOBAL_THREAD)
+	data->work.handler = kx022_work_cb;
+
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:164]: \-

File: drivers/sensor/kx022/kx022_trigger.c [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+	}
+	if (KX022_DEFAULT_INT_PIN == KX022_INT2) {
+		data->hw_tf->update_reg(data, KX022_REG_INC6, KX022_MAKS_INC6_DRDYI2,
+					KX022_INC6_DRDYI2);
+	}
+
+	kx022_mode(dev, KX022_OPERATING_MODE);
+}
+
+void kx022_Tilt_setup(const struct device *dev, sensor_trigger_handler_t handler)
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:220]: Just `kx022_tilt_setup`, maybe?

File: drivers/sensor/kx022/kx022_trigger.c [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+{
+	struct kx022_data *data = dev->data;
+
+	data->slope_handler = handler;
+
+	kx022_mode(dev, KX022_STANDY_MODE);
+
+	data->hw_tf->update_reg(data, KX022_REG_CNTL1, KX022_MASK_CNTL1_TPE, KX022_CNTL1_TPE_EN);
+
+	k_msleep(50);
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:230]: Magic number?

File: drivers/sensor/kx022/kx022_trigger.c [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+}
+
+void kx022_clear_setup(const struct device *dev)
+{
+	struct kx022_data *data = dev->data;
+
+	kx022_mode(dev, KX022_STANDY_MODE);
+	data->hw_tf->write_reg(data, KX022_REG_CNTL1, KX022_DEFAULT_CNTL1);
+	kx022_mode(dev, KX022_OPERATING_MODE);
+	k_msleep(50);
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:267]: Magic number?

File: drivers/sensor/kx022/kx022_trigger.c [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+		kx022_drdy_setup(dev, handler);
+		kx022_motion_setup(dev, handler);
+		kx022_Tilt_setup(dev, handler);
+		break;
+	case SENSOR_TRIG_NEAR_FAR:
+		kx022_clear_setup(dev);
+		kx022_Tilt_setup(dev, handler);
+		break;
+	default:
+		break;
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:295]: `-ENOTSUP`?

File: drivers/sensor/kx022/kx022_trigger.c [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+	k_msleep(50);
+}
+int kx022_trigger_set(const struct device *dev, const struct sensor_trigger *trig,
+		      sensor_trigger_handler_t handler)
+{
+	struct kx022_data *data = dev->data;
+	const struct kx022_config *cfg = dev->config;
+	uint8_t buf[6];
+
+	switch (trig->type) {
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:276]: Can do `kx022_mode(dev, KX022_STANDY_MODE);` before switch and `kx022_mode(dev, KX022_OPERATING_MODE);` before return or break.

File: samples/sensor/kx022/src/main.c [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+	printf("\t\tAccelerometer: Poling test Start\r\n");
+	do {
+		fetch_and_display(dev);
+		/* wait a while */
+		k_msleep(SLEEPTIME);
+
+		remaining_test_time -= SLEEPTIME;
+	} while (remaining_test_time > 0);
+}
+static void test_trigger_mode(const struct device *dev)
```
Comment (ycsin) [inline samples/sensor/kx022/src/main.c:196]: + 1 line

File: drivers/sensor/kx022/kx022_trigger.c [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+	kx022_handle_int(data->dev);
+}
+#endif
+
+/************************************************************
+ * info: kx022_mode use to set pc1 in standby or operating mode
+ * input: 0 - standby
+ *        1 - operating
+ * **********************************************************/
+
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:110]: \-

File: drivers/sensor/kx022/kx022_trigger.c [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+void kx022_mode(const struct device *dev, bool mode)
+{
+	struct kx022_data *data = dev->data;
+
+	data->hw_tf->update_reg(data, KX022_REG_CNTL1, KX022_MASK_CNTL1_PC1,
+				(mode << KX022_CNTL1_PC1_SHIFF));
+}
+/**************************************************************
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:118]: +1 line

File: drivers/sensor/kx022/kx022_trigger.c [hunk 17]
```diff
@@ -0,0 +1,315 @@
...
+#ifdef CONFIG_KX022_TRIGGER_GLOBAL_THREAD
+static void kx022_work_cb(struct k_work *work)
+{
+	struct kx022_data *data = CONTAINER_OF(work, struct kx022_data, work);
+
+	kx022_handle_int(data->dev);
+}
+#endif
+
+/************************************************************
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:105]: This comment block not quite inline with Zephyr's style

File: drivers/sensor/kx022/kx022_trigger.c (hunk 5 above)
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:145]: Use header defined value

File: drivers/sensor/kx022/kx022_trigger.c (hunk 7 above)
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:152]: default case likely not required

File: drivers/sensor/kx022/kx022_trigger.c (hunk 17 above)
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:105]: This comment block not quite inline with Zephyr's style

File: drivers/sensor/kx022/kx022_trigger.c (hunk 16 above)
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:118]: \+

File: drivers/sensor/kx022/kx022_trigger.c [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+		break;
+	case 2:
+		data->hw_tf->update_reg(data, KX022_REG_INC5, KX022_MAKS_INT2_EN, KX022_INT2_EN);
+		break;
+	default:
+		break;
+	}
+
+#if defined(CONFIG_KX022_TRIGGER_OWN_THREAD)
+	k_sem_init(&data->trig_sem, 0, UINT_MAX);
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:157]: `UINT_MAX` should be `K_SEM_MAX_LIMIT` now, according to recent Zephyr changes.

File: drivers/sensor/kx022/kx022_trigger.c [hunk 19]
```diff
@@ -0,0 +1,315 @@
...
+{
+	struct kx022_data *data = dev->data;
+
+	data->motion_handler = handler;
+
+	kx022_mode(dev, KX022_STANDY_MODE);
+	/* Enable Motion detection */
+	data->hw_tf->update_reg(data, KX022_REG_CNTL1, KX022_MASK_CNTL1_WUFE, KX022_CNTL1_WUFE);
+
+	data->hw_tf->update_reg(data, KX022_REG_ODCNTL, KX022_MASK_ODCNTL_OSA, KX022_ODCNTL_50HZ);
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:182]: Isn't this already configured in `kx022_init` of `kx022.c`?

File: drivers/sensor/kx022/kx022_trigger.c [hunk 20]
```diff
@@ -0,0 +1,315 @@
...
+}
+
+void kx022_drdy_setup(const struct device *dev, sensor_trigger_handler_t handler)
+{
+	struct kx022_data *data = dev->data;
+
+	data->drdy_handler = handler;
+	kx022_mode(dev, KX022_STANDY_MODE);
+	data->hw_tf->update_reg(data, KX022_REG_CNTL1, KX022_MASK_CNTL1_DRDYE, KX022_CNTL1_DRDYE);
+	data->hw_tf->update_reg(data, KX022_REG_ODCNTL, KX022_MASK_ODCNTL_OSA, KX022_ODCNTL_50HZ);
```
Comment (ycsin) [inline drivers/sensor/kx022/kx022_trigger.c:206]: Isn't this already configured in `kx022_init` of `kx022.c`?

File: dts/bindings/sensor/kionix,kx022.yaml [hunk 21]
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) 2021 G-Technologies Sdn. Bhd.
+# SPDX-License-Identifier: Apache-2.0
+
+description: KX022 16-Bit digital I2C Accelerometer sensor
+
+compatible: "kionix,kx022"
+
+include: i2c-device.yaml
+properties:
+    #  int1-gpios:
```
Comment (ycsin) [inline dts/bindings/sensor/kionix,kx022.yaml:10]: pending cleanup

File: samples/sensor/kx022/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,12 @@
+#
+#Copyright(c) 2021 G - Technologies Sdn.Bhd.
```
Comment (ycsin) [inline samples/sensor/kx022/CMakeLists.txt:2]: format

**Review by ycsin — CHANGES_REQUESTED** [review]
Potentially one of the most featured accelerometer driver in Zephyr, some comments for the driver part.

**Review by florasshi — COMMENTED** [review]
miss one line source "drivers/sensor/kx022/Kconfig" to drivers/sensor/Kconfig?

**Pull request closed without merging**